In [1]:
import gym
import random
import time
import numpy as np
from IPython.display import clear_output
env = gym.make("FrozenLake-v0").env
env.render()


SFFF
FHFH
FFFH
HFFG


In [2]:
action_space_size = env.action_space.n
state_space_size = env.observation_space.n
print(action_space_size)
print(state_space_size)

4
16


In [3]:
q_table = np.zeros((state_space_size, action_space_size)) #16 by 4
print(q_table)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


In [105]:
epochs = 100000
max_steps = 300

learning_rate = 0.09
discount_rate = 0.99 #works best

exploration_rate = 1
max_exploration_rate = 1
min_exploration_rate = 0.01
exploration_decay_rate = 0.001

In [106]:
rewards_all = []

for episode in range(epochs):
    state = env.reset()
    done = False
    current_reward = 0
    for step in range(max_steps):
        exploration_rate_threshold = random.uniform(0,1)
        if exploration_rate_threshold > exploration_rate:
            action = np.argmax(q_table[state,:])
        else:
            action = env.action_space.sample()
            
        new_state, reward, done, info = env.step(action)
        
        q_table[state, action] = q_table[state, action] * (1 - learning_rate) + learning_rate*(reward  + discount_rate * np.max(q_table[new_state,:])) 
        
        state = new_state
        current_reward += reward
        
        if done == True:
            break
        
    exploration_rate = min_exploration_rate + (max_exploration_rate - min_exploration_rate) * np.exp(-exploration_decay_rate*episode)
    
    rewards_all.append(current_reward)
    
rewards_print = np.split(np.array(rewards_all), epochs/500)

accuracy = []
for r in rewards_print:
    accuracy.append(sum(r/500))

In [107]:
count = 500
print("*******Average accuracy per 500 episodes******\n")
for i in accuracy:
    print(count, ": ", i)
    count += 500

*******Average accuracy per 500 episodes******

500 :  0.03800000000000002
1000 :  0.06200000000000004
1500 :  0.1580000000000001
2000 :  0.2480000000000002
2500 :  0.3760000000000003
3000 :  0.46800000000000036
3500 :  0.5640000000000004
4000 :  0.6120000000000004
4500 :  0.6740000000000005
5000 :  0.7040000000000005
5500 :  0.7400000000000005
6000 :  0.7160000000000005
6500 :  0.7460000000000006
7000 :  0.7340000000000005
7500 :  0.7680000000000006
8000 :  0.7140000000000005
8500 :  0.7060000000000005
9000 :  0.7340000000000005
9500 :  0.7660000000000006
10000 :  0.7400000000000005
10500 :  0.7500000000000006
11000 :  0.7420000000000005
11500 :  0.7940000000000006
12000 :  0.7560000000000006
12500 :  0.7520000000000006
13000 :  0.7700000000000006
13500 :  0.7900000000000006
14000 :  0.7560000000000006
14500 :  0.7820000000000006
15000 :  0.7260000000000005
15500 :  0.7720000000000006
16000 :  0.7680000000000006
16500 :  0.7440000000000005
17000 :  0.7140000000000005
17500 :  0.748000

In [109]:
print("max:", np.amax(accuracy), "at:", epochs/500 - (1+np.argmax(accuracy, axis=0)), "to last row")

max: 0.8180000000000006 at: 49.0 to last row


Trials:
* 0.806 & 0.802
    * epochs = 50000
    * max_steps = 200
    * learning_rate = 0.09
    * discount_rate = 0.99 
    * exploration_rate = 1
    * max_exploration_rate = 1
    * min_exploration_rate = 0.01
    * exploration_decay_rate = 0.001   
* 0.782 & 0.784
    * epochs = 10000(same as above but 1/5 the amount of epochs)
    * max_steps = 200
    * learning_rate = 0.09
    * discount_rate = 0.99 
    * exploration_rate = 1
    * max_exploration_rate = 1
    * min_exploration_rate = 0.01
    * exploration_decay_rate = 0.001 
* 0.818
    * epochs = 100000(same as above but 1/5 the amount of epochs)
    * max_steps = 300
    * learning_rate = 0.09
    * discount_rate = 0.99 
    * exploration_rate = 1
    * max_exploration_rate = 1
    * min_exploration_rate = 0.01
    * exploration_decay_rate = 0.001 


In [110]:
print(q_table)

[[0.5535096  0.52668216 0.51265377 0.50762933]
 [0.28948567 0.37899447 0.31819877 0.51529787]
 [0.43116976 0.43201604 0.42530365 0.47108634]
 [0.34345575 0.31659913 0.36544954 0.44183085]
 [0.56970152 0.45178143 0.44038537 0.39498596]
 [0.         0.         0.         0.        ]
 [0.34012094 0.15129203 0.13974951 0.13941627]
 [0.         0.         0.         0.        ]
 [0.22137729 0.39004322 0.33750436 0.59825085]
 [0.26988659 0.67030407 0.51635509 0.48457494]
 [0.59518243 0.42046194 0.35533805 0.22140881]
 [0.         0.         0.         0.        ]
 [0.         0.         0.         0.        ]
 [0.45462201 0.45246858 0.7733695  0.43959657]
 [0.70205455 0.91878884 0.76009179 0.74614123]
 [0.         0.         0.         0.        ]]


Testing the model

   -**to see the AI actually play, uncomment the print and time statements**

In [122]:
win_loss = []
from time import sleep
for episodes in range(1000):
    state = env.reset()
    done = False
    #time.sleep(2)
    while not done:
        action = np.argmax(q_table[state,:])
        new_state, reward, done, info = env.step(action)
        #clear_output(wait = True)
        #env.render()
        #sleep(.3)
        state = new_state
    if reward == 1:
        win_loss.append("Win")
        #print("******YOU WIN CONGRATZ!!!******")
        #time.sleep(1)
    else:
        win_loss.append("Loss")
        #print("******YOU LOSE******")
        #time.sleep(1)
    #sleep(2)
        

In [120]:
w = 0
l = 0
for i in win_loss:
    if i == "Win":
        w+=1
    else:
        l+=1
print("Wins:", w, "Losses:", l)

Wins: 1 Losses: 0
